### PR TITLE
GenBank: Don't discard 'linear' or 'circular' topology from LOCUS line

### PR DESCRIPTION
**Pull request by kblin**
Just ran into issue #363 myself, so here's a fix.

### CODE REVIEW

**Comment by kblin** [comment]
Ah, durn, BioSQL. I'll have a look.

File: Bio/SeqIO/InsdcIO.py [hunk 1]
```diff
@@ -604,6 +604,18 @@ def _get_data_division(self, record):
...
 
+    def _get_topology(self, record):
+        """Set the topology to 'circular', 'linear' if defined"""
+        max_topology_len = len('circular')
+
+        # return an empty placeholder string if not given
+        if 'topology' not in record.annotations:
+            return ' ' * max_topology_len
+
+        template = '%%-%ds' % max_topology_len
```
Comment (peterjc) [inline Bio/SeqIO/InsdcIO.py]: Would the string justification method be easier? i.e. ``.ljust(max_topology_len)``

File: Bio/SeqIO/InsdcIO.py [hunk 2]
```diff
@@ -647,17 +659,20 @@ def _write_the_first_line(self, record):
...
+        topology = self._get_topology(record)
+
         division = self._get_data_division(record)
 
         assert len(units) == 2
         assert len(division) == 3
         # TODO - date
         # TODO - mol_type
-        line = "LOCUS       %s %s %s    %s           %s %s\n" \
+        line = "LOCUS       %s %s %s    %s  %s %s %s\n" \
```
Comment (peterjc) [inline Bio/SeqIO/InsdcIO.py]: Why the double space before the topology? Which columns does the spec say are allocated here?

**Comment by peterjc** [comment]
I was wondering about tackling this for GenBank/EMBL input/output myself - I have some EMBL tweaking to do for work for an ENA submission, so should be in a position to actively test this next week :)

It was the BioSQL stuff that has held this back - really need to check where BioPerl (as the defacto BioSQL reference implementation) puts the topology field, but a short term hack would be to ignore the topology in the ``SeqRecord`` when writing to the database?

Also do we need ``melanin.gbk`` - lots of the existing ``Tests/GenBank/*.gb`` have useful ``LOCUS`` lines.

File: Bio/SeqIO/InsdcIO.py (hunk 1 above)
Comment (kblin) [inline Bio/SeqIO/InsdcIO.py]: Possibly. I can give that a spin.

File: Bio/SeqIO/InsdcIO.py (hunk 2 above)
Comment (kblin) [inline Bio/SeqIO/InsdcIO.py]: That's how I could make it align to the LOCUS line from the GenBank headers I looked at. I didn't find the exact definition for this in the spec. Then again, the spec is very confusing, which is the reason why we have that much fun with that format in the first place.

Comment (kblin) [inline Bio/SeqIO/InsdcIO.py]: Arguably, I could increase the `ljust()` for the molecule type by 1 as well instead.

**Comment by kblin** [comment]
I agree on the `melanin.gbk`. I had that lying around in my Tests/GenBank directory, so I just used it, and then later realized that it was missing. I'll refactor that to use another record.

File: Bio/SeqIO/InsdcIO.py (hunk 2 above)
Comment (peterjc) [inline Bio/SeqIO/InsdcIO.py]: That sounds sensible.

**Comment by peterjc** [comment]
Re BioSQL, as per my comments on #812, I have filed https://github.com/biosql/biosql/issues/5

**Comment by peterjc** [comment]
Rebased and with changes discussed above as https://github.com/peterjc/biopython/tree/topology - I hope to get the BioSQL side passing our tests next.

**Comment by peterjc** [comment]
Work in progress at https://github.com/biopython/biopython/pull/903 which includes the initial work from here by @kblin

**Pull request closed without merging**